react/src/util/opReturn.ts

```diff
@@ -1,20 +1,69 @@
-export const OP_RETURN_PREFIX = 'paybutton$';
+import { lib, enc } from 'crypto-js';
 
-export function stringToHex(str: string): string {
+//opReturn format is:
+//`04` - Pushdata of the Protocol Identifier, `50415900`
+//`50415900` - PayButton protocol identifier, ascii "PAY"
+//`00` - Version 0, pushed as `OP_0`
+//`09` - pushdata for the data payload, signifying this tx has 9 bytes of data
+//`0102030405060708aa` - data payload
+//`08` - pushdata for the optional nonce payload, signifying this tx has 8 bytes of nonce data
+//`0102030405060708` - The 8-byte nonce
+
+// All the below variables are already encoded to HEX
+export const OP_RETURN_PREFIX_PUSH_DATA = '04'; // \x04
+export const OP_RETURN_PREFIX = '50415900'; // PAY\x00
+export const VERSION = '00'; // \x00
+
+// 220 total bytes
+// - 5 from the 4-byte procol prefix with push data: '\x04PAY\x00'
+// - 1 for the version byte: '\x00'
+// - 9 from the 8-byte nonce with push data
+// = 205 available bytes
+export const BYTES_LIMIT = 220 - 5 - 1 - 9; // 205
+
+function stringToHex(str: string): string {
   return str
     .split('')
     .map(c => c.charCodeAt(0).toString(16).padStart(2, '0'))
     .join('');
 }
 
-export function parseOpReturn(
+function generatePushDataPrefixed8ByteNonce(): string {
+  // Generate 8 random bytes
+  const wordArray = lib.WordArray.random(8);
+
+  // Convert the word array to a hex string
+  const hexString = enc.Hex.stringify(wordArray);
+
+  // The result is 18 char long:
+  // ---
+  // a hex character encodes 4 bits of information (2⁴ = 16);
+  // ... therefore; 8 bytes = 72 bits => 72/4 = 16 hex chars
+  // + 1 byte of push data at the beggining = 2 hex chars
+  // = 18 chars
+  return `08${hexString}`;
+}
+
+export function parseOpReturnProps(
   opReturn: string | undefined,
 ): string | undefined {
-  if (opReturn === undefined) return undefined;
+  if (opReturn === undefined) {
+    opReturn = ''
+  }
   const bytesQuantity = new Blob([opReturn]).size;
 
-  if (bytesQuantity > 70) {
-    throw new Error(`Maximum 70 byte size exceeded: ${bytesQuantity}`);
+  if (bytesQuantity > BYTES_LIMIT) {
+    throw new Error(
+      `Maximum ${BYTES_LIMIT} byte size exceeded: ${bytesQuantity}`,
+    );
   }
-  return OP_RETURN_PREFIX + opReturn;
+  let pushData = bytesQuantity.toString(16).padStart(2, '0')
```

Comment:
This doesn't work for push datas > 75 bytes, so the above limit check makes no sense

Reply:
Why not?
![image](https://github.com/PayButton/paybutton/assets/21281174/eb92005c-ae16-4a98-9741-3b24e16bf7ca)

This doesn't work when bytesQuantity > 255, but the limit above makes it not a problem

Reply:
> Why not?

There are some obscure rules to how `script` works in bitcoin. To the point where it's probably worth using a `script` library, like the one in `bitcoinjs-lib` or BitGo's `utxo-lib`, to build any script strings.

https://en.bitcoin.it/wiki/Script

You can't push more than `0x4b` bytes with one byte of pushdata. `0x4c` is a special byte in script that means "the next byte is one byte of pushdata." `0x4d` is a special byte in script that means "the next two bytes are two bytes of pushdata."

imo your options here are

1. Do not use a library, but restrict pushes to between 1 and 75 bytes (seems like you are only coding for 2 pushes here that are always 8 or 9 bytes, so this is probably ok for now though not robust).
2. Use `bitcoinjs-lib` (or copy what Cashtab has done with `utxo-lib`), and have the pushdata created automatically with `script.push`

Reply:
Thanks for the info, this also answers the other question I had about why the need for two bytes of pushdata if it's always < 255.

But then I see two cases:
- data < 76 bytes; single byte for push data;
- 76 <= data < 205 bytes; two bytes for push data: `4c` + the actualy byte quantity

Since we don't have to care about bigger amounts than `255` (that would require more than two bytes of pushdata info, namely using the `4d` and `4e` op codes), this is still fairly simple to implement; I've made the changes to take that into account